### PR TITLE
Backports v0.18: PR's #14242, 15351 and 14519

### PR DESCRIPTION
**Pull request by pravblockc**
Backports v0.18: PR's #14242, 15351 and 14519

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
should drop 14242 (it's empty) + see below (feels like quite a few earlier PRs are missing)

**Comment by pravblockc** [comment]
> should drop 14242 (it's empty) + see below (feels like quite a few earlier PRs are missing)

Okay, I will take-up earlier PR's that are mandatory to port this(will fix this later)

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -321,7 +349,15 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
...
-                        "ttt        (numeric) The number of seconds that the server has been running\n"
+                "{\n"
+                " \"active_commands\" (array) All active commands\n"
+                "  [\n"
+                "   {               (object) Information about an active command\n"
+                "    \"method\"       (string)  The name of the RPC command \n"
+                "    \"duration\"     (numeric)  The running time in microseconds\n"
+                "   },...\n"
+                "  ]\n"
+                "}\n"
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply src/rpc/server.cpp]: did drop this PR as few more dependent PR's need to be ported before this

File: doc/developer-notes.md [hunk 2]
```diff
@@ -269,6 +270,51 @@ make cov
...
+trade-offs](https://lwn.net/Articles/420403/) of setting these kernel
+parameters.
+
+To profile a running bitcoind process for 60 seconds, you could use an
+invocation of `perf record` like this:
+
+```sh
+$ perf record \
+    -g --call-graph dwarf --per-thread -F 140 \
+    -p `pgrep bitcoind` -- sleep 60
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply doc/developer-notes.md]: done

File: test/functional/test_framework/test_node.py [hunk 3]
```diff
@@ -17,6 +17,9 @@
 import tempfile
 import time
 import urllib.parse
+import collections
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply test/functional/test_framework/test_node.py]: done

File: test/functional/test_framework/test_node.py [hunk 4]
```diff
@@ -280,6 +299,111 @@ def assert_debug_log(self, expected_msgs):
...
+            return
+
+        perc_increase_memory_usage = (after_memory_usage / before_memory_usage) - 1
+
+        if perc_increase_memory_usage > increase_allowed:
+            self._raise_assertion_error(
+                "Memory usage increased over threshold of {:.3f}% from {} to {} ({:.3f}%)".format(
+                    increase_allowed * 100, before_memory_usage, after_memory_usage,
+                    perc_increase_memory_usage * 100))
+
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply test/functional/test_framework/test_node.py]: updated

File: src/rpc/server.cpp [hunk 5]
```diff
@@ -330,14 +366,41 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
...
- * Call Table
- */
+static UniValue getrpcinfo(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() > 0)
+        throw std::runtime_error(
+                "getrpcinfo\n"
+                    "\nReturns details of the RPC server.\n"
+                    "\nResult:\n"
+                    "ttt        (numeric) The number of seconds that the server has been running\n"
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply src/rpc/server.cpp]: Removed 14982 as it got few other PR's to be back ported

File: test/lint/lint-python-dead-code.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (pravblockc) [reply test/lint/lint-python-dead-code.sh:1]: back ported other PR and updated/rebased changes

**Comment by pravblockc** [comment]
> should drop 14242 (it's empty) + see below (feels like quite a few earlier PRs are missing)

added commit that was missing for 14242

File: test/functional/test_framework/address.py [hunk 7]
```diff
@@ -41,14 +41,6 @@ def scripthash_to_p2sh(hash, main = False):
...
     return byte_to_base58(hash, version)
 
-def key_to_p2pkh(key, main = False):
-    key = check_key(key)
-    return keyhash_to_p2pkh(hash160(key), main)
-
-def script_to_p2sh(script, main = False):
-    script = check_script(script)
-    return scripthash_to_p2sh(hash160(script), main)
-
```
Comment (UdjinM6) [inline test/functional/test_framework/address.py:51]: 14519: this should be reverted

File: src/crc32c/.ycm_extra_conf.py [hunk 8]
```diff
@@ -122,21 +122,3 @@ def FlagsForClangComplete(file_path, build_root):
...
-  file_path = os.path.realpath(filename)
-
-  flags = BASE_FLAGS
-  clang_flags = FlagsForClangComplete(file_path, build_root)
-  if clang_flags:
-    flags += clang_flags
-
-  final_flags = MakeRelativePathsInFlagsAbsolute(flags, build_root)
-
-  return {'flags': final_flags}
```
Comment (UdjinM6) [inline src/crc32c/.ycm_extra_conf.py:142]: 14519: this should be reverted

File: test/lint/lint-python-dead-code.sh [hunk 9]
```diff
@@ -15,5 +15,5 @@ fi
 
 vulture \
     --min-confidence 60 \
-    --ignore-names "argtypes,connection_lost,connection_made,converter,data_received,daemon,errcheck,get_ecdh_key,get_privkey,is_compressed,is_fullyvalid,msg_generic,on_*,optionxform,restype,set_privkey,*serialize_v2" \
-    $(git ls-files -- "*.py" ":(exclude)contrib/" ":(exclude)src/crc32c/" ":(exclude)test/functional/test_framework/address.py")
+    --ignore-names "argtypes,connection_lost,connection_made,converter,data_received,daemon,errcheck,get_ecdh_key,get_privkey,is_compressed,is_fullyvalid,msg_generic,on_*,optionxform,restype,set_privkey,profile_with_perf" \
+    $(git ls-files -- "*.py" ":(exclude)contrib/" ":(exclude)test/functional/data/invalid_txs.py")
```
Comment (UdjinM6) [inline test/lint/lint-python-dead-code.sh:19]: 14519: should not add unrelated changes here
```suggestion
    --ignore-names "argtypes,connection_lost,connection_made,converter,data_received,daemon,errcheck,get_ecdh_key,get_privkey,is_compressed,is_fullyvalid,msg_generic,on_*,optionxform,restype,set_privkey,*serialize_v2,profile_with_perf" \
    $(git ls-files -- "*.py" ":(exclude)contrib/" ":(exclude)src/crc32c/" ":(exclude)test/functional/test_framework/address.py")
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Pls see below + should drop an empty `test/sanitizer_suppressions/ubsan` file from 14242

File: test/functional/test_framework/address.py (hunk 7 above)
Comment (pravblockc) [reply test/functional/test_framework/address.py:51]: @UdjinM6  Could you please confirm whether issues at https://gitlab.com/dashpay/dash/-/jobs/1508393724 need to be considered(any errors on running python linter). That is the reason I have removed some methods. Please confirm

Comment (UdjinM6) [reply test/functional/test_framework/address.py:51]: Linter fails there because 1. you broke `test/lint/lint-python-dead-code.sh` (see https://github.com/dashpay/dash/pull/4329#discussion_r690690560 below) and 2. there is an issue fixed by #4344

**Comment by UdjinM6** [comment]
Pls rebase on top of `develop` and add missing changes to `test/sanitizer_suppressions/ubsan` (14242), this file should exist now that we merged 14764 via #4341.

**Comment by pravblockc** [comment]
> Pls rebase on top of `develop` and add missing changes to `test/sanitizer_suppressions/ubsan` (14242), this file should exist now that we merged 14764 via #4341.

above mentioned changes has been done

**Review by UdjinM6 — APPROVED** [review]
👍 

utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit
